### PR TITLE
Support for Categorical features in CalculateFeatureContribution of LightGBM

### LINKED ISSUE #3272: Exception on using LightGBM trainer with FeatureContributionCalculation and OneHotEncoding
### System information

- **OS version/distro**: Windows 10
- **.NET Version (eg., dotnet --info)**: 4.7.1

### Issue

- **What did you do?**
I used FeatureContributionCalculation with LightGbm trainer. My data pipeline contains OneHotEncoding features.
- **What happened?**
When I try to get feature contribution calculation I get the following exception
```
System.InvalidOperationException
  HResult=0x80131509
  Message=Splitter/consolidator worker encountered exception while consuming source data
  Source=Microsoft.ML.Data
  StackTrace:
   at Microsoft.ML.Data.DataViewUtils.Splitter.Batch.SetAll(OutPipe[] pipes)
   at Microsoft.ML.Data.DataViewUtils.Splitter.Cursor.MoveNextCore()
   at Microsoft.ML.Data.RootCursorBase.MoveNext()
   at Microsoft.ML.Data.ColumnCursorExtensions.<GetColumnArrayDirect>d__3`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at ConsoleApp1.Program.BuildTrainEvaluateAndSaveModel(MLContext mlContext) in C:\Users\vladimir.kuryshev\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 150
   at ConsoleApp1.Program.Main(String[] args) in C:\Users\vladimir.kuryshev\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 45

Inner Exception 1:
ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: slot

   at Microsoft.ML.Data.VBuffer`1.GetItemOrDefault(Int32 slot)
   at Microsoft.ML.Trainers.FastTree.InternalRegressionTree.AppendFeatureContributions(VBuffer`1& src, BufferBuilder`1 contributions)
   at Microsoft.ML.Trainers.FastTree.InternalTreeEnsemble.GetFeatureContributions(VBuffer`1& features, VBuffer`1& contribs, BufferBuilder`1& builder)
   at Microsoft.ML.Trainers.FastTree.TreeEnsembleModelParameters.<>c__DisplayClass30_0`2.<Microsoft.ML.Model.IFeatureContributionMapper.GetFeatureContributionMapper>b__0(VBuffer`1& src, VBuffer`1& dst)
   at Microsoft.ML.Data.DataViewUtils.Splitter.InPipe.Impl`1.Fill()
   at Microsoft.ML.Data.DataViewUtils.Splitter.<>c__DisplayClass5_1.<ConsolidateCore>b__2()
```
### Source code / logs
My code example
```
   IDataView trainingDataView = mlContext.Data.LoadFromTextFile<TaxiTrip>(TrainDataPath, hasHeader: true, separatorChar: ',');
   var dataProcessPipeline = mlContext.Transforms.CopyColumns(outputColumnName: DefaultColumnNames.Label, inputColumnName: nameof(TaxiTrip.FareAmount))
                            .Append(mlContext.Transforms.Categorical.OneHotEncoding(outputColumnName: VendorIdEncoded, inputColumnName: nameof(TaxiTrip.VendorId)))
                            .Append(mlContext.Transforms.Categorical.OneHotEncoding(outputColumnName: RateCodeEncoded, inputColumnName: nameof(TaxiTrip.RateCode)))
                            .Append(mlContext.Transforms.Categorical.OneHotEncoding(outputColumnName: PaymentTypeEncoded, inputColumnName: nameof(TaxiTrip.PaymentType)))
                            .Append(mlContext.Transforms.Normalize(outputColumnName: nameof(TaxiTrip.PassengerCount), mode: NormalizingEstimator.NormalizerMode.MeanVariance))
                            .Append(mlContext.Transforms.Normalize(outputColumnName: nameof(TaxiTrip.TripTime), mode: NormalizingEstimator.NormalizerMode.MeanVariance))
                            .Append(mlContext.Transforms.Normalize(outputColumnName: nameof(TaxiTrip.TripDistance), mode: NormalizingEstimator.NormalizerMode.MeanVariance))
                            .Append(mlContext.Transforms.Concatenate(DefaultColumnNames.Features, VendorIdEncoded, RateCodeEncoded, PaymentTypeEncoded, nameof(TaxiTrip.PassengerCount)
                            , nameof(TaxiTrip.TripTime), nameof(TaxiTrip.TripDistance)));

   var trainer = mlContext.Regression.Trainers.LightGbm(labelColumnName: DefaultColumnNames.Label, featureColumnName: DefaultColumnNames.Features);
   var trainingPipeline = dataProcessPipeline.Append(trainer);

   var trainedModel = trainingPipeline.Fit(trainingDataView);
   IDataView predictions = trainedModel.Transform(testDataView);

   var featureContributionCalculation = mlContext.Model.Explainability.FeatureContributionCalculation(trainedModel.LastTransformer.Model);
   var featureContributionData = featureContributionCalculation.Fit(predictions).Transform(predictions);
   var contributions = featureContributionData.GetColumn<float[]>(mlContext, DefaultColumnNames.FeatureContributions).ToList();
```
I used standard data from "taxi-fare-train.csv" file from ML.Net examples.

- **Note** Issue can't be reproduced If i remove OneHotEncoding from pipeline or change trainer to FastTree for example.

### PR DESCRIPTION
**Pull request by antoniovs1029**
Fixes #3272

As explained [here](https://github.com/dotnet/machinelearning/issues/3272#issuecomment-611849910), `CalculateFeatureContribution` would throw an exception when used on LightGBM models that had `UseCategoricalSplit` enabled, because there was no support to calculate feature contribution for categorical features. Here I add that support, and one test to replicate the original scenario were an exception was thrown.

### CODE REVIEW

File: src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs [hunk 1]
```diff
@@ -1515,28 +1515,59 @@ public void AppendFeatureContributions(in VBuffer<float> src, BufferBuilder<floa
...
+                    }
+
+                    // What if we went the other way?
+                    var ghostLeaf = GetLeafFrom(in src, otherWay);
+                    var ghostOutput = GetOutput(ghostLeaf);
+
+                    // If the ghost got a smaller output, the contribution of the categorical features is positive, so
+                    // the contribution is true minus ghost.
+                    foreach(var ifeat in CategoricalSplitFeatures[node])
+                        contributions.AddFeature(ifeat, (float)(trueOutput - ghostOutput));
```
Comment (antoniovs1029) [inline src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs:1543]: Code-wise, I guess this is the correct way to find which features are involved in the categorical split of this node (as it is done in a similar way [here](https://github.com/dotnet/machinelearning/blob/1ea53362a692e551737341787d7b49541c28dfb8/src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs#L851-L877))

But I don't know if this is the "mathematically correct" way of calculating feature contributions for categorical features.

**Comment by rauhs** [comment]
Does this also affect #3766 ?

**Comment by antoniovs1029** [comment]
Hi, @rauhs. This PR here won't affect issue #3766. Nonetheless, having a quick look at that issue, it seems it is similar in nature to issue #3272 which is fixed in this PR.
Issue #3766 is obtaining the Gains from the following code, which is obtaining the feature's indices from SplitFeatures[node]. 

[https://github.com/dotnet/machinelearning/blob/8660ecc7742a02ac896dc77c93c67c366ff4647e/src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs#L1334-L1335](https://github.com/dotnet/machinelearning/blob/8660ecc7742a02ac896dc77c93c67c366ff4647e/src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs#L1334-L1335)

As explained [here](https://github.com/dotnet/machinelearning/issues/3272#issuecomment-611849910) the `SplitFeatures[]` array has "-1" for categorical splits, and it shouldn't be used for such splits (`CategoricalSplitFeatures[][]` should be used instead). So to solve issue #3766 we should also add code to support categorical features when calculating the GainMap. Problem is that I don't know what's the "mathematically correct" way to do it. I will ask around to see if I can get the correct way to do it, and add that code as well.

Thanks for pointing to that issue!

---
In reply to: [612639332](https://github.com/dotnet/machinelearning/pull/5018#issuecomment-612639332) [](ancestors = 612639332)

File: src/Microsoft.ML.FastTree/RegressionTree.cs [hunk 2]
```diff
@@ -57,7 +57,7 @@ public abstract class RegressionTreeBase
         /// (2) the categorical features indexed by <see cref="GetCategoricalCategoricalSplitFeatureRangeAt(int)"/>'s
         /// returned value with nodeIndex=i is NOT a sub-set of <see cref="GetCategoricalSplitFeaturesAt(int)"/> with
         /// nodeIndex=i.
-        /// Note that the case (1) happens only when <see cref="CategoricalSplitFlags"/>[i] is true and otherwise (2)
+        /// Note that the case (1) happens only when <see cref="CategoricalSplitFlags"/>[i] is false and otherwise (2)
```
Comment (antoniovs1029) [inline src/Microsoft.ML.FastTree/RegressionTree.cs:60]: It looks to me this doc was wrong, as it's inconsistent with what these other lines below say:

https://github.com/dotnet/machinelearning/blob/8660ecc7742a02ac896dc77c93c67c366ff4647e/src/Microsoft.ML.FastTree/RegressionTree.cs#L76-L80

https://github.com/dotnet/machinelearning/blob/8660ecc7742a02ac896dc77c93c67c366ff4647e/src/Microsoft.ML.FastTree/RegressionTree.cs#L100-L106

File: src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs (hunk 1 above)
Comment (antoniovs1029) [reply src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs]: So I've updated this. I am still not sure the updated version is correct. But I think it's the closest to how feature contribution is calculated for numerical feature splits, and considering that for other other cases (FastTree, Gam, etc...) categorical features are treated the same as any other feature (ignoring the fact they're categorical) when calculating feature contribution.

**Pull request merged**